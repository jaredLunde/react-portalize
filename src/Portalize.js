import React from 'react'
import ReactDOM from 'react-dom'
import PropTypes from 'prop-types'


/**
<Portalize container='#portals'>
  <div>Inject me</div>
</Portalize>
*/
export const PORTALS = {}

export default class Portalize extends React.Component {
  static propTypes = {
    container: PropTypes.string.isRequired
  }

  static defaultProps = {
    container: '#portals'
  }

  nodes = null

  render () {
    this.nodes =
      typeof document !== 'undefined' && document.querySelectorAll(this.props.container)

    if (this.nodes === false) {
      PORTALS[this.props.container] = this.props.children
      return null
    }
    else if (this.nodes.length === 0) {
      return null
    }
    else {
      const portals = []

      for (let i = 0; i < this.nodes.length; i++) {
        portals.push(ReactDOM.createPortal(this.props.children, this.nodes[i]))
      }

      return portals
    }
  }
}
