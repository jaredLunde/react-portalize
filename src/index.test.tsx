/* jest */
import * as React from 'react'
import {render} from '@testing-library/react'
import Portalize, {PORTALS} from './index'

describe('<Portalize>', () => {
  it('should render into #portals by default', () => {
    const portalRoot = document.createElement('div')
    portalRoot.setAttribute('id', 'portals')
    document.body.appendChild(portalRoot)

    const result = render(
      <Portalize>
        <div>Hello world</div>
      </Portalize>
    )

    expect(result.baseElement).toMatchSnapshot()
    expect(Object.values(PORTALS).length).toBe(0)
    document.body.removeChild(portalRoot)
  })

  it('should render into a portal by custom selector', () => {
    const portalRoot = document.createElement('div')
    portalRoot.setAttribute('class', 'portals')
    document.body.appendChild(portalRoot)

    const result = render(
      <Portalize container=".portals">
        <div>Hello world</div>
      </Portalize>
    )

    expect(result.baseElement).toMatchSnapshot()
    expect(Object.values(PORTALS).length).toBe(0)
    document.body.removeChild(portalRoot)
  })

  it('should not render into a non-existent selector', () => {
    const portalRoot = document.createElement('div')
    portalRoot.setAttribute('class', 'blah')
    document.body.appendChild(portalRoot)

    const result = render(
      <Portalize container=".portals">
        <div>Hello world</div>
      </Portalize>
    )

    expect(result.baseElement).toMatchSnapshot()
    document.body.removeChild(portalRoot)
  })
})
