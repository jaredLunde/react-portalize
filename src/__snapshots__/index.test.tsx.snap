// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Portalize> should not render into a non-existent selector 1`] = `
<body>
  <div
    class="blah"
  />
  <div />
</body>
`;

exports[`<Portalize> should render into #portals by default 1`] = `
<body>
  <div
    id="portals"
  >
    <div>
      Hello world
    </div>
  </div>
  <div />
</body>
`;

exports[`<Portalize> should render into a portal by custom selector 1`] = `
<body>
  <div
    class="portals"
  >
    <div>
      Hello world
    </div>
  </div>
  <div />
</body>
`;
